Define table specifications

tcompa opened the pull request:
Close #529
Close #587 
Close #588

## Checklist before merging
- [x] I added an appropriate entry to `CHANGELOG.md`

jluethi commented on docs/roi_tables.md:
````diff
@@ -0,0 +1,333 @@
+}
+```
+
+#### Single table (default)
````
Single table default vs. segmented objects?
What is default? ROI?

jluethi commented on docs/roi_tables.md:
```diff
@@ -0,0 +1,333 @@
+## Scope
+
+In `fractal-tasks-core`, regions of interest (ROIs) are three-dimensional
+regions of space delimited by orthogonal planes. ROI tables are stored as
```
What does "delimited by orthogonal planes" mean? Can we rephrase this?

jluethi commented on docs/roi_tables.md:
```diff
@@ -0,0 +1,333 @@
+We have several use cases for tables:
+
+1. We keep track of the positions of the Field of Views (FOVs) within a well, after stitching the corresponding FOV images into a single whole-well array.
+2. We keep track of the original state before some transformations are applied - e.g. shifting FOVs to avoid overlaps, or shifting a multiplexing cycle during registration.
```
We also use ROI tables to apply the actual registration in case of multiplexing

jluethi replied:
At least when the registration is a translation, that's supported

jluethi commented on docs/roi_tables.md:
```diff
@@ -0,0 +1,333 @@
+
+#### Single table (segmented objects)
+
+When each table row corresponds to (the bounding box of) a segmented object,
```
I'm confused now. Aren't the ROI tables the bounding box tables? Is this a feature table (as e.g. coming from napari workflow wrapper?) or a masked ROI table (e.g. cellpose table output?

jluethi commented on docs/roi_tables.md:
```diff
@@ -0,0 +1,333 @@
+the following six columns:
+
+* `x_micrometer`, `y_micrometer`: the lower bounds of the XY intervals defining the ROI, in micrometers;
+* `z_micrometer`: the lower bound of the Z interval defining the ROI, in arbitrary units or in micrometers;
```
Oh, I was not aware we're not using the actual units here. That would be a bug, the intent is to use the actual units!

jluethi commented on docs/roi_tables.md:
```diff
@@ -0,0 +1,333 @@
+ROI tables may also include other optional columns:
+
+* `x_micrometer_original` and `y_micrometer_original`, which are a copy of `x_micrometer` and `y_micrometer` taken before applying some transformation;
+* `label`, which is used within measurement tables as a reference to the labels corresponding to a row of measurements (see [description of `instance_key` above](#single-table-segmented-objects)).
```
also used within masking ROI tables, right?

jluethi commented on docs/roi_tables.md:
```diff
@@ -0,0 +1,333 @@
+> Notes:
+>
+> 1. The **axes origin** for the ROI positions (e.g. for `x_micrometer`) is set
+>    to coincide with the top-left corner of a well (for the YX axes) and with
```
Top left corner of an image
=> Our spec should apply to any OME-NGFF image, not only to plates with wells

jluethi reviewed (commented):
Great first version @tcompa 
Let's discuss at our call. A few bigger points:

Is it the doc for ROI tables or tables in general?
Title is ROI, but text says general table spec. Let's briefly discuss at our call
In short, I see 2 equivalent use-cases: 
- ROIs (and sub-cases): Each row is a region of interest
- Measurements: Each row is a measured object (e.g. related to a given label in the label image)


If our z_micrometer are not used in micrometers, but as arbitrary units: That's actually a bug. We just never noticed, because many users image with z pixel size of 1. But some users image with 0.6 or 2.0

jluethi commented:
Outlook: Axis generalization

jluethi commented:
Link to PR for fusing FOVs to single well: 
https://github.com/ome/ngff/pull/137

Nomenclature:
Tiled (instead of stitched or fused)

jluethi commented:
TODO: Mention masked loading functionality

jluethi commented:
3 tables:
Core Tables
ROI tables
Advanced ROI tables (=> naming TBD)

jluethi commented on docs/tables.md:
```diff
@@ -0,0 +1,458 @@
+tasks) always include two specific ROI tables:
+
+* The table named `well_ROI_table`, which covers the NGFF image corresponding to the whole well[^1];
+* The table named `FOV_ROI_table`, which lists all original FOVs.
```
lists all original field of views (FVOs)

jluethi commented on docs/tables.md:
```diff
@@ -0,0 +1,458 @@
+* The table named `well_ROI_table`, which covers the NGFF image corresponding to the whole well[^1];
+* The table named `FOV_ROI_table`, which lists all original FOVs.
+
+Each one of these two tables includes ROIs that are only defined in the XY
```
Still not clear about this. They are defined in Z, just as spanning the whole Z range, no?

jluethi replied:
The spec should be that they define the XYZ extent of those FOVs (and the well and FOV always span the whole Z range). But as far as the ROI table is concerned, it only knows that because we put in the correct len_z_micrometer)

Whether our code contains an assumption somewhere that those pixel sizes in Z would always be 1: to be verified, but that would be a bug

jluethi commented on docs/tables.md:
```diff
@@ -0,0 +1,458 @@
+  flexible, as it can be used to process e.g. a whole well, a set of FOVs, or a
+  set of custom regions of the array.
+
+### Reading/writing tables
```
Love that you included code examples for reading tables!! ❤️

jluethi replied:
Can we add a few lines like:

```
# To convert a table to pandas:
df = table.to_df()
print(df)
```

jluethi commented on docs/tables.md:
```diff
@@ -0,0 +1,458 @@
+
+To read an `AnnData` table from a Zarr group, one may use the [`read_zarr`
+function](https://anndata.readthedocs.io/en/latest/generated/anndata.read_zarr.html).
+In the following example a NGFF image was created by sticthing together two
```
stitching

jluethi commented on docs/tables.md:
```diff
@@ -0,0 +1,458 @@
+To read an `AnnData` table from a Zarr group, one may use the [`read_zarr`
+function](https://anndata.readthedocs.io/en/latest/generated/anndata.read_zarr.html).
+In the following example a NGFF image was created by sticthing together two
+field of views, where each one is made of a stack of five Z planes.
```
Stack of five Z planes with 1um spacing between the planes

jluethi commented on docs/tables.md:
```diff
@@ -0,0 +1,458 @@
+  `x_micrometer`), in favor of a more general definition of units.
+* The `z_micrometer` and `len_z_micrometer` columns are currently required in
+  all ROI tables, even when the ROIs actually define a two-dimensional XY
+  region; in that case, we set `z_micrometer=0` and `len_z_micrometer` is such
```
`len_z_micrometer` = 1? 
Or what do we set?

jluethi commented on docs/tables.md:
```diff
@@ -0,0 +1,458 @@
+* We may re-evaluate whether `AnnData` tables are the most appropriate tool. For
+  the record, Zarr does not natively support storage of dataframes (see e.g.
+  https://github.com/zarr-developers/numcodecs/issues/452), which is one
+  aspect in favor of sticking with the `anndata` library.
```
Our plan is to follow whatever serialised table specification becomes part of the OME-NGFF standard.

jluethi reviewed (approved):
Looks great, I really like the reworked version @tcompa . Great job!

Only open remaining question is about wording on whether Z is used for well & FOV ROIS, let's briefly discuss tomorrow. The rest are minor wording suggestions or typos :)

tcompa commented:
Thanks for the review.
I addressed all comments but one via https://github.com/fractal-analytics-platform/fractal-tasks-core/pull/582/commits/43a622e7afd52d3849a9fa7809208a0b7d390526, feel free to double-check.
